terraform/transform_config.go

```diff
@@ -3,121 +3,94 @@ package terraform
 import (
 	"errors"
 	"fmt"
+	"log"
 
-	"github.com/hashicorp/go-multierror"
-	"github.com/hashicorp/terraform/config"
 	"github.com/hashicorp/terraform/config/module"
+	"github.com/hashicorp/terraform/dag"
 )
 
-// ConfigTransformer is a GraphTransformer that adds the configuration
-// to the graph. The module used to configure this transformer must be
-// the root module. We'll look up the child module by the Path in the
-// Graph.
+// ConfigTransformer is a GraphTransformer that adds all the resources
+// from the configuration to the graph.
+//
+// The module used to configure this transformer must be the root module.
+//
+// Only resources are added to the graph. Variables, outputs, and
+// providers must be added via other transforms.
+//
+// Unlike ConfigTransformerOld, this transformer creates a graph with
+// all resources including module resources, rather than creating module
+// nodes that are then "flattened".
```

Comment:
Yes to both, but there is no flattening operation.

A history on flattening: Terraform used to treat modules as sub-graphs (that were never flattened) and when you "enter" the subgraph you'd built a new `EvalContext` and when you exited you'd throw it away. This was how modules worked in TF 0.3 (first version with modules I think).

We soon realized with bug reports that modules are not an isolated graph. There are complex ordering requirements that require entering/exiting the module boundary multiple times. The simplest example is:

  1. module input `A` created by `root.foo` resource
  2. module resource `module.bar` created with input `A` 
  3. module output `X` created with a value from `module.bar`
  4. `root.baz` resource created with module output `X`
  5. module input `B` uses value from resource `root.baz`
  6. module resource `module.qux` uses input `B`

Notice how this requires in/out, in/out at least twice. Hence, the isolated subgraph didn't work. This was a serious serious bug that made modules basically unusable in Terraform 0.3. We needed a quick bolt-on solution to fix this in TF 0.3.1. That solution ended up being "well, we already have these correct subgraphs, what if we just flattened them in and added the necessary extra edges?" And so we did. 

And graph flattening has been with us since.

But it is complicated. We knew even in 0.3.1 the right solution would've just been to put ALL resources in ONE graph from the beginning. There is no "flatten" step: you just have all resources with the proper address from the beginning.

This transform does that. Gone with flattening.